### ISSUE TITLE
Explore Performance Optimizations for NNBenchmark

### ISSUE THREAD

**Issue by tuzhucheng** [tuzhucheng/nd4j-conv-benchmarks#1]
This repo contains similar code as those used for Deeplearning4J performance tests in https://arxiv.org/abs/1707.08275, minus boilerplate code and dependencies of the [Anserini](https://github.com/castorini/Anserini) project.

To run the test once inside the `java` directory and after compiling the project (`mvn compile`), do:
```
./run.sh run-nn-test 500 100 50 20 5 4 4 201
```

This is what each number above means: `./run.sh run-nn-test [repeat] [numFilters] [embeddingH] [embeddingW] [filterW] [padding] [numExternalFeatures] [numHiddenLayerUnits]`. These configurations are close to the ones used on the [TrecQA](https://github.com/castorini/data/tree/master/TrecQA) dataset in the paper.

Currently the implementation of the neural network is here: https://github.com/tuzhucheng/nd4j-conv-benchmarks/blob/594149f2fc028d4c8c0300a39d6167bb8face5c0/java/src/main/java/com/example/app/SiameseNN.java.

Any suggestions on how to improve the performance are welcome. Currently on the list are:
+ Try this on the latest Deeplearning4J snapshot release rather than 0.8.0 found in the maven repo.
+ Use Deeplearning4J pooling rather than current implementation which uses nd4j directly.

**Comment by crockpotveggies** [comment]
Thank you very much for putting up this repo. We did something similar over here on the Skymind side and are currently looking into this. Comparing our benchmarks to yours will also be helpful.

I do know that @raver119 reviewed your implementation and has already discovered a couple optimizations with how you're implementing your network. We're going to have a proper response for you next week.

**Comment by crockpotveggies** [comment]
Alright sent an email with further details, but first step here is we need to compare apples to apples. In some cases, DL4J experiences a 700% performance boost when compiled with MKL. There's a clear difference.

Regarding your implementation, @raver119 has made some comments available here: https://gist.github.com/raver119/e46f8156aa5a0749dbc008ba7ac4d713

I strongly suggest using 0.8.1-SNAPSHOT since that is our 0.9.0 which will be coming out very soon. I also added notes from raver below:

> In this particular case, this architecture should be implemented via regular ComputationGraph in DL4j, so it will benefit from recent improvements like workspaces etc.

> As for pooling layer: it was improved recently, and became single native op used from dl4j.
As for allocations: we have workspaces now, and dl4j uses that feature a lot internally. In this case it probably should be just WorkspaceMode.SINGLE

> And obvious question. I’ve inspected pytorch source code. Their convolution/pooling implementation is equal to dl4j, with the only difference - dl4j uses multithreading via openmp right in im2col, and  pytorch - doesn’t.

> And one more thing: compilation difference. We’re not aware of pytorch installation method. If that was built from sources etc - it might be optimized for «current cpu architecture», like libnd4j does via «-a native» compilation flag. But DL4j available via maven (which was used there), is compiled for generic x86_64. This difference is worth of another 5-10%.

Instructions for compiling libnd4j from scratch: https://deeplearning4j.org/buildinglocally. Happy to help there, we also have a build script in the main DL4J repo that handles some of the monotonous tasks.

**Comment by raver119** [comment]
Also please note - MK. Since your network is relatively small one, MKL use makes dramatic difference here.
